packages/apps/src/app.ts

```diff
@@ -366,6 +359,44 @@ export class App {
     return res;
   }
 
+  /**
+   * Refresh the tokens for the app
+   */
+  protected async refreshTokens() {
+    await this.refreshBotToken();
+    await this.refreshGraphToken();
+  }
+
+  private async refreshBotToken(force = false) {
+    if (this.credentials) {
+      // Only do it if the token isn't there, or if it's expired, or if force is true
+      if (
+        !this._tokens.bot ||
+        (this._tokens.bot.expiration != null && this._tokens.bot.expiration < Date.now()) ||
+        force
+      ) {
+        this.log.debug('Refreshing bot token');
+        const botResponse = await this.api.bots.token.get(this.credentials);
+        this._tokens.bot = new JsonWebToken(botResponse.access_token);
+      }
+    }
+  }
+
+  private async refreshGraphToken(force = false) {
+    if (this.credentials) {
+      // Only do it if the token isn't there, or if it's expired, or if force is true
+      if (
+        !this._tokens.graph ||
+        (this._tokens.graph.expiration != null && this._tokens.graph.expiration < Date.now()) ||
```

Comment:
Should we add a threshold? Like maybe 20ms?

Reply:
I would go as far as 5 minutes, just to give a ton of buffer for long running requests.

Reply:
Added a 5 min buffer